File: src/main/java/com/learning/java/inheritance/AbstractClassSample.java
package com.learning.java.inheritance;

/*
 * Java: абстрактный (abstract) класс
 */


abstract class Printer {
	public abstract void print(); // абстрактный метод - прнуждает определять собственную реализацию в своих классах-наследниках
	public void status() {
		System.out.println("Printer is running");
	}
}

class ColorPrinter extends Printer {
	@Override
	public void print() {		
		System.out.println("Print in color");
	}
}


public class AbstractClassSample {

    public static void main(String[] args) {
        ColorPrinter cp = new ColorPrinter();
        cp.status();
        cp.print();
    }
}

File: src/main/scala/com/learning/scala/inheritance/AbstractClassSample.scala
package com.learning.scala.inheritance

/*
 * Scala: абстрактный (abstract) класс
 */


/* ( Абстрактный класс - такой abstract-класс нельзя создавать, он обязательно должен иметь своего класса-наследника.. ) */
abstract class Printer {
  def print: Unit /* в Scala, метод без определения своего тела - указывает такой метод является абстрактным, и класс-наследник вынужден определить собственную реализацию для этого абстрактного метода */
  def status: Unit = println("Printer is running")
}

/* Класс-наследник */
class ColorPrinter extends Printer {
  def print: Unit = println("Print in color")  // при определении тела для абстрактного метода - ключевое слово 'override' НЕтребуется
}


object AbstractClassSample {

  def main(args: Array[String]) {
    val cp = new ColorPrinter
    cp status;
    cp print;
  }
}
